pkg/kubectl/cmd/apply/prune/prune.go

```diff
@@ -0,0 +1,230 @@
+	pruneGvks := [][]string{
+		{"", "v1", "ConfigMap"},
+		{"", "v1", "Endpoints"},
+		{"", "v1", "Namespace"},
+		{"", "v1", "PersistentVolumeClaim"},
+		{"", "v1", "PersistentVolume"},
+		{"", "v1", "Pod"},
+		{"", "v1", "ReplicationController"},
+		{"", "v1", "Secret"},
+		{"", "v1", "Service"},
+		{"batch", "v1", "Job"},
+		{"extensions", "v1beta1", "DaemonSet"},
+		{"extensions", "v1beta1", "Deployment"},
+		{"extensions", "v1beta1", "Ingress"},
+		{"extensions", "v1beta1", "ReplicaSet"},
+		{"apps", "v1beta1", "StatefulSet"},
+		{"apps", "v1beta1", "Deployment"},
+	}
+
+	// If group/version/kind specified in "prune-whitelist" flag, then use those values
+	if len(gvks) != 0 {
+		var err error
+		pruneGvks, err = parsePruneWhitelist(gvks)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// Parse groupVersionKind from passed prune-whitelist flag values
+	var groupVersionKinds []*meta.RESTMapping
+	for _, gvk := range pruneGvks {
+		mapping, err := mapper.RESTMapping(schema.GroupKind{Group: gvk[0], Kind: gvk[2]}, gvk[1])
+		if err != nil {
+			return nil, err
+		}
+		groupVersionKinds = append(groupVersionKinds, mapping)
+	}
+
+	return groupVersionKinds, nil
+}
+
+// Example gvks: "core/v1/endpoints", "apps/v1beta1/deployment"
+// Returns validated gvks as array of arrays of strings
+func parsePruneWhitelist(gvks []string) ([][]string, error) {
+
+	var parsedGvks [][]string
+	for _, groupVersionKind := range gvks {
+		gvk := strings.Split(groupVersionKind, "/")
+		if len(gvk) != 3 {
+			return nil, fmt.Errorf("invalid GroupVersionKind format: %v, please follow <group/version/kind>", groupVersionKind)
+		}
+		if gvk[0] == "core" {
+			gvk[0] = ""
+		}
+		parsedGvks = append(parsedGvks, gvk)
+	}
+	return parsedGvks, nil
+}
+
+// Prune resources not previously visited.
+func (p *Pruner) Prune(visitedNamespaces sets.String, visitedUids sets.String) error {
+	// Prune namespaced resources
+	for n := range visitedNamespaces {
+		for _, m := range p.GroupVersionKinds {
+			if m.Scope.Name() == meta.RESTScopeNameNamespace {
+				if err := p.pruneMapping(n, m, visitedUids); err != nil {
+					return fmt.Errorf("error pruning namespaced object %v: %v", m.GroupVersionKind, err)
+				}
+			}
+		}
+	}
+	// Prune non-namespaced resources
+	for _, m := range p.GroupVersionKinds {
+		if m.Scope.Name() == meta.RESTScopeNameRoot {
+			if err := p.pruneMapping(metav1.NamespaceNone, m, visitedUids); err != nil {
+				return fmt.Errorf("error pruning non-namespaced object %v: %v", m.GroupVersionKind, err)
+			}
+		}
+	}
+	return nil
+}
+
+// Deletes objects in namespace/group/version/kind which have not previously been visited.
+func (p *Pruner) pruneMapping(namespace string, mapping *meta.RESTMapping, visitedUids sets.String) error {
+	c, err := p.ClientFunc(mapping)
+	if err != nil {
+		return err
+	}
+
+	export := false
+	apiVersion := mapping.GroupVersionKind.Version
+	objList, err := resource.NewHelper(c, mapping).List(
```

Comment:
Wrap this in a type that performs some of the filterings - e.g. skip if no apply annotation